src/drivers/esp32/src/esp_slip.c

```diff
@@ -0,0 +1,315 @@
+/**
+ * ,---------,       ____  _ __
+ * |  ,-^-,  |      / __ )(_) /_______________ _____  ___
+ * | (  O  ) |     / __  / / __/ ___/ ___/ __ `/_  / / _ \
+ * | / ,--´  |    / /_/ / / /_/ /__/ /  / /_/ / / /_/  __/
+ *    +------`   /_____/_/\__/\___/_/   \__,_/ /___/\___/
+ *
+ * Crazyflie control firmware
+ *
+ * Copyright (C) 2021 Bitcraze AB
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, in version 3.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program. If not, see <http://www.gnu.org/licenses/>.
+ *
+ * @file esp_slip.c
+ * Protocol for assembling, sending, receiving and decoding SLIP packets to/from the ESP32 ROM bootloader
+ *  
+ */
+
+#include <string.h>
+
+#include "esp_slip.h"
+#include "uart2.h"
+
+#define ESP_OVERHEAD_LEN 8
+
+typedef enum
+{
+  receiveStart,
+  receiveDirection,
+  receiveCommand,
+  receiveSize,
+  receiveValue,
+  receiveData,
+  receiveEnd,
+  error,
+} ESPblReceiveState;
+static ESPblReceiveState espblReceiveState = receiveStart;
+
+static uint8_t slipSize[2];
+static uint8_t slipValue[4];
+static uint8_t slipDataIndex = 0;
+static uint8_t slipSizeIndex = 0;
+static uint8_t slipValueIndex = 0;
+static bool inEscapeSequence = false;
+
+static uint32_t sendSize;
+
+static uint8_t generateChecksum(uint8_t *sendBuffer, esp_slip_send_packet *senderPacket)
+{
+  uint8_t checksum = 0xEF; // seed
+  for (int i = 0; i < senderPacket->dataSize - 16; i++)
+  {
+    checksum ^= sendBuffer[9 + 16 + i];
+  }
+  return checksum;
+}
+
+static void sendSlipPacket(uint32_t size, uint8_t *data, coms_sendbuffer_t sendBufferFn)
+{
+  uint32_t i;
+  static uint8_t dmaSendBuffer[UART2_DMA_BUFFER_SIZE];
+  uint32_t dmaSendSize = 0;
+
+  for (i = 0; i < size; i++)
+  {
+    if ((data[i] == 0xC0 && i != 0 && i != size - 1) || (data[i] == 0xDB && i != 0 && i != size - 1))
```

Comment:
Consider creating constants for magic numbers, especially if used multiple times, for instance 0xC0